tests/agent_server/deploy/test_scheduler_agent.py

```diff
@@ -71,14 +72,53 @@ async def open_version(self, version: int) -> None:
     async def close_version(self, version: int) -> None:
         pass
 
+    async def stop(self) -> None:
+        pass
+
     async def join(self) -> None:
         pass
 
 
+class ManagedExecutor(DummyExecutor):
+    """
+    Dummy executor that can be driven explicitly by a test case.
+
+    Executor behavior must be controlled through the `deploys` property. It exposes a mapping from resource ids
+    to futures. Simply set the desired outcome as the result on the appropriate future.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._deploys: dict[ResourceIdStr, asyncio.Future[const.ResourceState]] = {}
+
+    @property
+    def deploys(self) -> Mapping[ResourceIdStr, asyncio.Future[const.ResourceState]]:
+        return self._deploys
+
+    async def stop(self) -> None:
+        # resolve hanging futures to prevent test hanging during teardown
+        for deploy in self._deploys.values():
+            deploy.set_result(const.ResourceState.undefined)
+
+    async def execute(self, gid: uuid.UUID, resource_details: ResourceDetails, reason: str) -> const.ResourceState:
+        assert resource_details.rid not in self._deploys
+        self._deploys[resource_details.rid] = asyncio.Future()
```

Comment:
didn't know you could do this, I expect to have to use the create_future of the ioloop.

Reply:
You can do this, but I checked the docs just now and it turns out that they recommend your approach: "the recommended way to create a Future object is to call [loop.create_future()](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_future)".

Reply:
```suggestion
        self._deploys[resource_details.rid] = asyncio.get_running_loop().create_future()
```